### PR TITLE
feat: Fetch settings and divide events view model

### LINKED ISSUE #2083: Break events view model into smaller ones
**Would you like to work on the issue?**

yes

### LINKED ISSUE #2101: Fetch timer from settings
- Timer for ticket registration should be according to the organizer preference.
- Message for the timer is displaying wrong. 

**Would you like to work on the issue?**
<!-- Please let us know if you can work on it or the issue should be assigned to someone else. --yes

### PR DESCRIPTION
**Pull request by liveHarshit**
Fixes #2101 
Fixes #2083 

<p align="center">
 
</p>

### CODE REVIEW

File: app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt [hunk 1]
```diff
@@ -297,7 +295,7 @@ class AttendeeFragment : Fragment(), ComplexBackPressFragment {
         rootView.timeoutInfoTextView.text =
             getString(R.string.ticket_timeout_info_message, event.orderExpiryTime.toString())
 
-        val timeLeft: Long = if (attendeeViewModel.timeout == -1L) COUNT_DOWN_TIME * 60 * 1000L
+        val timeLeft: Long = if (attendeeViewModel.timeout == -1L) event.orderExpiryTime * 60 * 1000L
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: This was moved to settings, so it won't be available in event

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: But how it is working currently. I've tested it, frontend is also using the same endpoint.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Working on dev or prod?

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: https://github.com/fossasia/open-event-server/pull/6083

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: > Working on dev or prod?

production, while creating an event, time is asking form the organizer.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Yeah, so again, this was moved to settings, so it won't be available in event. (10 days ago)

Always consider dev to be latest. It'll break when server is updated.

That's why it's asked for client devs to only use dev instance. You must be ready for any breaking change being applied on server. If you'll use prod environment, how will you deal with new changes?

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: > Always consider dev to be latest. It'll break when server is updated.

Okay, I should also check with the dev instance. It was during creating the meetup event. Will update.

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Only admin can reterive order expiry time - 
image

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Open an issue to allow clients to see order expiry time. If this is true, then I guess timer on frontend is wrong as well

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Will fetch setting and test +1
Should I fetch setting at the starting of the app and store into the database or fetch every time when required?

**Comment by liveHarshit** [comment]
https://github.com/fossasia/open-event-server/issues/6185

File: app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt (hunk 1 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Save into DB and when retrieving, first load from DB and then send a request to refresh

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: > Save into DB and when retrieving, first load from DB and then send a request to refresh

One problem will there if I load time from the database first and after that refresh it from the server then count down timer will start again by observing new value.

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Currently, I'm directly fetching settings at the time of ordering tickets. Updated.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: You will not show it again, it'll be shown next time user opens the order screen. If even you are directly fetching, save in DB

**Comment by liveHarshit** [comment]
Updated

File: app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt [hunk 2]
```diff
@@ -260,9 +258,13 @@ class AttendeeFragment : Fragment(), ComplexBackPressFragment {
             .observe(viewLifecycleOwner, Observer {
                 loadEventDetailsUI(it)
                 setupPaymentOptions(it)
-                if (attendeeViewModel.pendingOrder.value != null) {
-                    setupCountDownTimer(it)
-                }
+            })
+
+        attendeeViewModel.getSettings()
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: There is no guarantee that settings have been fetched till the next statement is executed

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Then onErrorResumeNext will execute and settings will fetch from api.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Still there is no guarantee that settings have been fetched till the next statement is executed, as getsettings is asynchronous

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: Sorry, I don't understand which next statement. During fetching/getting settings progress dialog is displaying and if settings fetch successfully or there is an error(=15), value for mutableOrderExpiryTime is set and only then countdown timer will start.

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: The statement which is immediately proceeding getSettings

File: app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeViewModel.kt [hunk 3]
```diff
@@ -391,6 +395,21 @@ class AttendeeViewModel(
...
+        compositeDisposable += settingsService.getSettings()
+            .withDefaultSchedulers()
+            .doOnSubscribe {
+                mutableProgress.value = true
+            }.doFinally {
+                mutableProgress.value = false
+            }.subscribe({
+                mutableOrderExpiryTime.value = it.orderExpiryTime
+            }, {
+                mutableOrderExpiryTime.value = 15
```
Comment (anhanh11001) [inline app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeViewModel.kt]: use const value here

**Comment by liveHarshit** [comment]
Updated

**Comment by iamareebjamal** [comment]
Approving but there is a need of polishing

**Pull request merged**